[collectd 6] Use real collectd metric functions in gpu_sysman plugin tests

eero-t opened the pull request:
ChangeLog: Use real collectd metric functions in "gpu_sysman" plugin tests

There have been lot of changes in metric handling code, so it's better to use real metric functions rather than mocked ones.

This allows removing the mock functions, but makes it much easier for unrelated changes to break `gpu_sysman_test.c` compilation (e.g .utf-8 support already broke earlier version of this).

eero-t commented:
@octo CI fails to unrelated error:
```
  Submodule 'gnulib' (git://git.savannah.gnu.org/gnulib.git) registered for path 'gnulib'
  Submodule 'opentelemetry-proto' (https://github.com/open-telemetry/opentelemetry-proto/) registered for path 'opentelemetry-proto'
  Cloning into '/__w/collectd/collectd/gnulib'...
  Cloning into '/__w/collectd/collectd/opentelemetry-proto'...
  Error: fatal: read error: Connection reset by peer
  Fetched in submodule path 'gnulib', but it did not contain 2f8140bc8ce5501e31dcc665b42b5df64f84c20c. Direct fetching of that commit failed.
  Error: The process '/usr/bin/git' failed with exit code 1
```

eero-t commented:
> @octo CI fails to unrelated error on "debian10" (although "debian11" passes):

Oddly, everything passed in https://github.com/collectd/collectd/pull/4177 that was rebased on top of this. Maybe that error was just transitory network error.

octo commented on src/gpu_sysman_test.c:
```diff
@@ -68,6 +68,25 @@
 
+/* utilities needed from collectd core */
+#include <unistd.h>
+/* common.c fails to -Werror=unused-parameter, so
```
Can you post the error you're getting? This sounds like it would be easy to fix.

eero-t replied:
`getpwnam_r()` implementation in `common.c` does not use all of its parameters.  I worked around that by using `#define HAVE_GETPWNAM_R 1`.

After that, and fixing the includes, there are still following warnings:
```
gcc -Werror -Wall -Wextra -Wpedantic -Wcast-align=strict -Wformat-security -Wnull-dereference -Wstrict-overflow=4 -Warray-bounds=2 -Wno-aggressive-loop-optimizations -I. -Idaemon -O3 -g -D_FORTIFY_SOURCE=2 gpu_sysman_test.c -o gpu_sysman_test
In file included from gpu_sysman_test.c:74:
utils/common/common.c: In function ‘check_create_dir’:
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 256 equals destination size [-Werror=stringop-truncation]
   97 |   strncpy(dest, src, n);
      |   ^
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 256 equals destination size [-Werror=stringop-truncation]
utils/common/common.c: In function ‘notification_init’:
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 256 equals destination size [-Werror=stringop-truncation]
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 128 equals destination size [-Werror=stringop-truncation]
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 128 equals destination size [-Werror=stringop-truncation]
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 128 equals destination size [-Werror=stringop-truncation]
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 128 equals destination size [-Werror=stringop-truncation]
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 128 equals destination size [-Werror=stringop-truncation]
utils/common/common.c: In function ‘notification_init_metric’:
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 256 equals destination size [-Werror=stringop-truncation]
utils/common/common.c: In function ‘walk_directory’:
utils/common/common.c:1142:5: error: assuming signed overflow does not occur when changing X +- C1 cmp C2 to X cmp C2 -+ C1 [-Werror=strict-overflow]
 1142 | int walk_directory(const char *dir, dirwalk_callback_f callback,
      |     ^~~~~~~~~~~~~~
```

octo replied:
Which version of GCC is that? I have been unable to reproduce with GCC on Debian 12:

```
$ make V=1 src/utils/common/common.lo
depbase=`echo src/utils/common/common.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
/bin/bash ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I./src  -I./src -I./src/daemon -DPREFIX='"/tmp/collectd"' -DCONFIGFILE='"/tmp/collectd/etc/collectd.conf"' -DLOCALSTATEDIR='"/tmp/collectd/var"' -DPKGLOCALSTATEDIR='"/tmp/collectd/var/lib/collectd"' -DPLUGINDIR='"/tmp/collectd/lib/collectd"' -DPKGDATADIR='"/tmp/collectd/share/collectd"'   -Wall -Werror -Werror -Wall -Wextra -Wpedantic -Wcast-align=strict -Wformat-security -Wnull-dereference -Wstrict-overflow=4 -Warray-bounds=2 -Wno-aggressive-loop-optimizations -MT src/utils/common/common.lo -MD -MP -MF $depbase.Tpo -c -o src/utils/common/common.lo src/utils/common/common.c &&\
mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I./src -I./src -I./src/daemon -DPREFIX=\"/tmp/collectd\" -DCONFIGFILE=\"/tmp/collectd/etc/collectd.conf\" -DLOCALSTATEDIR=\"/tmp/collectd/var\" -DPKGLOCALSTATEDIR=\"/tmp/collectd/var/lib/collectd\" -DPLUGINDIR=\"/tmp/collectd/lib/collectd\" -DPKGDATADIR=\"/tmp/collectd/share/collectd\" -Wall -Werror -Werror -Wall -Wextra -Wpedantic -Wcast-align=strict -Wformat-security -Wnull-dereference -Wstrict-overflow=4 -Warray-bounds=2 -Wno-aggressive-loop-optimizations -MT src/utils/common/common.lo -MD -MP -MF src/utils/common/.deps/common.Tpo -c src/utils/common/common.c  -fPIC -DPIC -o src/utils/common/.libs/common.o
$ gcc --version
gcc (Debian 12.2.0-14) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

octo replied:
#4233 attempts to fix this.

eero-t replied:
This was with the  default GCC version in Ubuntu 22.04 LTS:
`gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)`

However, to get GCC to actually analyze the code more thoroughly, you also need to use high optimization levels (I used `-O3`, your command line does not seem to  specify any optimizations?).

And your command line was missing `-D_FORTIFY_SOURCE=2`: https://wiki.ubuntu.com/Security/Features#Built_with_Fortify_Source

octo commented on src/gpu_sysman_test.c:
```diff
@@ -68,6 +68,25 @@
+/* common.c fails to -Werror=unused-parameter, so
+ * string helpers still need mocked versions.
+ */
+#if 0
+#include <dirent.h>
+#include <math.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+/* comment to avoid contrib/format.sh breaking code */
+#include "utils/common/common.c"
+#endif
```
Is this left-over from development or the preferred solution should we be able to resolve the issues with `common.c`?

eero-t replied:
I added commit fixing includes in `common.c` & `strbuf.c`, and removed these from the test code.

octo commented on src/gpu_sysman_test.c:
```diff
@@ -842,120 +861,6 @@ int plugin_dispatch_metric_family(metric_family_t const *fam) {
-}
-
 /* ------------------------------------------------------------------------- */
 /* mock up of collectd plugin API */
```
Would it make sense to move this to `src/daemon/plugin_mock.c`?

eero-t replied:
I need the information passed to that API, to validate correct working of the plugin's config and metric handling. What is checked from metric and logging parts in the mocked functions is specific to sysman plugin, config options less so, but I think it makes sense to keep it all plugin specific.

octo replied:
Sure, works for me.

octo commented:
> > @octo CI fails to unrelated error on "debian10" (although "debian11" passes):
> 
> Oddly, everything passed in #4177 that was rebased on top of this. Maybe that error was just transitory network error.

git.savannah.gnu.org had transient errors all day. The "Actions" tab has a "re-run failed workflows" button that is quite useful in that situation.

octo reviewed (approved):
LGTM

eero-t commented:
Rebased to to latest `collectd-6.0`, fixed new/remaining GCC warnings in `collectd` core code, and ordered commits so that every one of them compiles with stricter compiler options.

(Sorry for the extra pushes, annoyingly my local and CI `clang-format` disagree on include order.)

----

@octo I had to downgrade one warning in my own tests due to this:
```
$ gcc -v
gcc version 13.2.1 20231011 (Red Hat 13.2.1-4) (GCC)

$ gcc -O2 -I src -I src/daemon -DFP_LAYOUT_NEED_NOTHING=1 -Werror -Wstrict-overflow=4 src/utils/common/common.c 
src/utils/common/common.c: In function ‘walk_directory’:
src/utils/common/common.c:1146:5: error: assuming signed overflow does not occur when changing X +- C1 cmp C2 to X cmp C2 -+ C1 [-Werror=strict-overflow]
 1146 | int walk_directory(const char *dir, dirwalk_callback_f callback,
      |     ^~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```
`-Wstrict-overflow` takes values from 1-5, and any value above 2 gives above warning when GCC optimization level is >1.

I do not see what GCC could be complaining about `common.c` though.

eero-t commented:
I hadn't understood how collectd code handles system includes (= nearly all code relies on them being included from collectd.h, but that happens only when HAVE_CONFIG_H is defined, i.e. configure has generated config.h etc).

=> Reverted my include addition commits, cleaned up those from test-code, and documented how tests should be run manually.

octo commented on src/gpu_sysman_test.c:
```diff
@@ -64,9 +69,15 @@
-#define KERNEL_LINUX 1
-#define FP_LAYOUT_NEED_NOTHING 1
-#include "gpu_sysman.c" /* test this */
+/* utilities needed from collectd core */
+#include "daemon/metric.c"
+#include "daemon/utils_cache_mock.c"
+#include "utils/common/common.c"
+#include "utils/metadata/meta_data.c"
+#include "utils/strbuf/strbuf.c"
+#include "utils/utf8/utf8.c"
```
Why doesn't `gpu_sysman.c` include these?

eero-t replied:
Because `gpu_sysman.c` is a collect plugin [1], whereas  `gpu_sysman_test.c` is a separate binary.

[1] i.e. collectd provides the symbols:
```
$ strip -s collectd
$ readelf -sW collectd | grep -v UND | awk '/FUNC *GLOBAL/{print $8}' | cut -b-7 | sort | uniq -c | sort -nr
     58 plugin_
     19 meta_da
     14 metric_
     13 uc_meta
     12 uc_get_
      9 strbuf_
...
```

The reasons why test binary includes all the necessary code directly, instead of linking to archives built during collectd build, are several:
* With all used code being visible to GCC, it can do proper whole-program analysis
* I'm testing the code in question with more strict options i.e. I will improve quality also in them
* I run these tests separately from the collectd build

octo replied:
Oh! That is unexpected. I'd actually much prefer to consistently build tests by linking to modules that are not under test.

* Whole program analysis: Why is this needed? My understanding is that whole program analysis is primarily done to optimize the code, and I don't particularly care about the performance of unit tests.
* Stricter cflags: We can add an additional build to the CI system that builds everything with stricter flags. We can probably also run tests with ASAN instead of Valgrind.
* Alternative build method: I'm not keen on maintaining alternative, bespoke way of building parts of the tree. In other words, maintainability is more important to me than supporting such one-off use cases.

eero-t replied:
> * Whole program analysis: Why is this needed? My understanding is that whole program analysis is primarily done to optimize the code, and I don't particularly care about the performance of unit tests.

Warnings analysis is part of compiler optimization passes.  If you do not enable optimizations, code checks can miss a lot of things.  And with contents of all functions visible to compiler, it can do better checks (along with those optimizations), for things actually used in the test.


> * Stricter cflags: We can add an additional build to the CI system that builds everything with stricter flags.

Enabling all the flags I'm interested about for whole collectd, would require a lot of fixes to rest of the code first.  Now I've needed to fix only things that are included directly to the tester.

(I can work on fixing those for collectd core, but I do not feel comfortable doing it for plugins that I'm not using, unless my changes are really trivial/obvious improvements.)

> We can probably also run tests with ASAN instead of Valgrind.

Yes, that would be really nice!


> * Alternative build method: I'm not keen on maintaining alternative, bespoke way of building parts of the tree. 

Sorry, I forgot to mention most important reason of including just the needed C-files instead of using the archives.

To do the checks requires by tests, I need to intercept (have special versions of) few collectd functions:
* metric submit: what metrics were reported or not reported, and with what exact values?
* logging: were warnings/errors generated as expected, or unexpectedly?
* config: were config option parsing results expected?

Unless you have a good suggestion on how to get access to submitted metrics and logs when using real collectd functions (without needing to implement both write & log plugins functionality), I think using build-internal archives would mess the tester.

> In other words, maintainability is more important to me than supporting such one-off use cases.

Yes that's understandable.

Skipping this PR is one option (for avoiding direct inclusion potentially causing later test breakages with unrelated collectd changes).

I would then need to update the mocked metric functions (removed in this PR) instead though.

octo replied:
I thought about this a bit, because I believe that we need to improve our test coverage, particularly for frequently used plugins. My rough idea so far was:

* (Optional) Rename `plugin_mock.c` to `plugin_fake.c`
* In `src/daemon/plugin_mock.c` write a test-only implementation of `plugin_dispatch_metric_family()` that clones all metric families and stores the copy somewhere.
* Add functions for retrieving and resetting the stored metric families. Make these functions accessible to tests via `src/daemon/plugin_mock.h`.

This should allow (read) plugins to actually call the dispatch function successfully. Once the read function returns the test can inspect what metric families were dispatched, if any.

Possible alternatives:

* Implement a fake `plugin_register_write()` so that the fake `plugin_dispatch_metric_family()` call write callbacks without all the thread start-up required by the real `plugin.c`. Then provide a "write plugin for testing" implementation with the "copy and store in memory" behavior outlines above.
* Provide a test harness to do the above, but with the "real" `plugin.c` implementation. This would probably give the best test coverage, but is also the most complex to pull off.

We already have a fake logging implementation, we'd just need to also store log messages to make them available to the tests. This should be rather easy.

If the plugin is written so that it doesn't so any I/O in the config callback, the config parsing should be easy to test. 
#4238 is doing it, although with a temporary file for now. If we add the ability to parse a string to the config file code, then we should be able to test this without I/O also.

eero-t replied:
**Config**

Sysman plugin test code calls registered config function for each key.   If real plugin code is used, it could call plugin config function directly instead.

**Logging**

Logs should not be much of a problem either.  The main thing (besides logs being printed for debugging) is able to get counts of things logged above given level, and being able to reset those counts.  With that, test can verify that failing things fail (with log output), and ones that should not fail, don't do that.

**Metrics**

This is the issue. Propsed thing will make debugging metric issues much harder.   My own fake code uses asserts on all possible errors, instead of just logging something, or exiting, like collectd code does.  Sysman plugin tester does a *lot* of things and backtraces from asserts right in the submit path, give a huge help for tracking down issues.  Backtraces do not help if metrics are collected first, and only then verified.

=> Better would be to have a test callback for metric submit

eero-t replied:
I could avoid those includes by linking to `libcommon.la` and adding couple of dummy functions (for things that requires, but does not link).

Those includes are still in the code, but behind `BUILD_STANDALONE` ifdef, so that my own testing can do whole program analysis.

octo commented on src/gpu_sysman_test.c:
```diff
@@ -10,14 +10,21 @@
+ * - Fedora: sudo dnf install oneapi-level-zero-devel
+ * - Debian: sudo apt install libze-dev
+ *
+ * Building unit-tests (after running "configure", so that
+ * "config.h" + other generated includes are present):
+ *	gcc -I. -Idaemon -DHAVE_CONFIG_H -O3 -g --coverage -Werror \
+ *	 -Wall -Wextra -Wpedantic -Wcast-align=strict -Wformat-security \
+ *	 -Wnull-dereference -Wstrict-overflow=2 -Warray-bounds=2 \
+ *	 -D_FORTIFY_SOURCE=2 -Wno-aggressive-loop-optimizations \
+ *	 gpu_sysman_test.c -o test_plugin_gpu_sysman
```
Why do you need to document this?

The tests should automatically be built and run as part of `make check`. If you need to build this test individuall, run `make test_plugin_gpu_sysman`. If you want to run only this test, you can use `make test_plugin_gpu_sysman.log` (only really useful during development).

eero-t replied:
> Why do you need to document this?

* Comment documents everything that I'm testing with that file (used compiler options are just one of those items)
* It lists (stricter) compiler options that included collectd core files should IMHO pass (reason for the previous fix commits to them)

> The tests should automatically be built and run as part of `make check`. If you need to build this test individually, run `make test_plugin_gpu_sysman`. If you want to run only this test, you can use `make test_plugin_gpu_sysman.log` (only really useful during development).

Thanks!

Often I will (build and) run it separately though, because I want that to be done under Valgrind (maybe later compile tests with ASAN too), and I'm interested also in test coverage:
```
Optimized build coverage info:

File 'gpu_sysman_test.c'
Lines executed:86.50% of 511

File 'gpu_sysman.c'
Lines executed:87.46% of 1132

File 'daemon/metric.c'
Lines executed:27.10% of 369

File 'utils/strbuf/strbuf.c'
Lines executed:0.00% of 126

File 'utils/utf8/utf8.c'
Lines executed:71.43% of 7

File 'utils/metadata/meta_data.c'
Lines executed:1.29% of 309

File 'utils/common/common.c'
Lines executed:3.41% of 499

File 'daemon/utils_cache_mock.c'
Lines executed:0.00% of 19
```

(As you can see, some of the includes are needed just duo to another collectd core file referring symbols in them, nothing is actually used from them during Sysman plugin tests.)

eero-t replied:
Documented this a bit better in the comment.

eero-t commented:
Current version is failing because `config.h` apparently gets configured with HAVE_CAPABILITY in some CI targets, and therefore:
```
/__w/collectd/collectd/src/utils/common/common.c:1485: undefined reference to `cap_get_bound'
/usr/bin/ld: /__w/collectd/collectd/src/utils/common/common.c:1488: undefined reference to `cap_get_proc'
/usr/bin/ld: /__w/collectd/collectd/src/utils/common/common.c:1493: undefined reference to `cap_get_flag'
/usr/bin/ld: /__w/collectd/collectd/src/utils/common/common.c:1498: undefined reference to `cap_free'
/usr/bin/ld: /__w/collectd/collectd/src/utils/common/common.c:1495: undefined reference to `cap_free'
```

=> Trying to include real `*.c` files directly is too much whack-a-mole game.

I guess I'll next need to try how badly testing breaks if tester is instead linked with `libplugin_mock.la`, and calls Sysman plugin's plugin-API functions directly.

Looking at `Makefile.am`, there does not seem to be any archive that I could use to link all relevant real functions, as closest to that seemed to be what `collectd` daemon itself links.

octo commented:
> Looking at `Makefile.am`, there does not seem to be any archive that I could use to link all relevant real functions, as closest to that seemed to be what `collectd` daemon itself links.

That is was libplugin_mock.la needs to become IMHO – a drop-in replacement for the daemon to test plugins. That's for example why I recently threw out the mocked/faked "cache" implementation and linked with the real one instead. By linking with the "real" implementation tests can now cover plugins that require the cache to function.

eero-t commented:
@octo This works OK again, and your feedback should be handled.

As to my removal of (apparently redundant) `-lm` lines from `Makefile.am`, should I split that to separate PR?

eero-t commented:
Dropped the extra `Makefile.am` change, rebased to latest & squashed commits to a single one.  Will merge this later today (based on the earlier approval,  unless there are comments/objections/CI-check fails).
